src/query/service/src/pipelines/processors/transforms/hash_join/result_blocks.rs

```diff
@@ -270,6 +281,118 @@ impl JoinHashTable {
         Ok(())
     }
 
+    fn create_marker_block(&self, has_null: bool, markers: Vec<MarkerKind>) -> Result<DataBlock> {
+        let mut validity = MutableBitmap::with_capacity(markers.len());
+        let mut boolean_bit_map = MutableBitmap::with_capacity(markers.len());
+
+        for m in markers {
+            let marker = if m == MarkerKind::False && has_null {
+                MarkerKind::Null
+            } else {
+                m
+            };
+            if marker == MarkerKind::Null {
+                validity.push(false);
+            } else {
+                validity.push(true);
+            }
+            if marker == MarkerKind::True {
+                boolean_bit_map.push(true);
+            } else {
+                boolean_bit_map.push(false);
+            }
+        }
+        let boolean_column = BooleanColumn::from_arrow_data(boolean_bit_map.into());
+        let marker_column = Self::set_validity(&boolean_column.arc(), &validity.into())?;
+        let marker_schema = DataSchema::new(vec![DataField::new(
+            &self
+                .hash_join_desc
+                .marker_join_desc
+                .marker_index
+                .ok_or_else(|| ErrorCode::LogicalError("Invalid mark join"))?
+                .to_string(),
+            NullableType::new_impl(BooleanType::new_impl()),
+        )]);
+        Ok(DataBlock::create(DataSchemaRef::from(marker_schema), vec![
+            marker_column,
+        ]))
+    }
+
+    fn mark_join_probe_subquery<Key, IT>(
+        &self,
+        hash_table: &HashMap<Key, Vec<RowPtr>>,
+        probe_state: &mut ProbeState,
+        keys_iter: IT,
+        input: &DataBlock,
+    ) -> Result<DataBlock>
+    where
+        Key: HashTableKeyable + Clone + 'static,
+        IT: Iterator<Item = Key> + TrustedLen,
+    {
+        let has_null = {
+            let has_null_ref = self.hash_join_desc.marker_join_desc.has_null.read();
+            *has_null_ref
+        };
+
+        let mut markers = vec![MarkerKind::False; input.num_rows()];
+        let valids = &probe_state.valids;
+        if self.hash_join_desc.other_predicate.is_none() {
+            // todo(youngsofun): can be optimized as semi-join?
```

Comment:
I'll do an optimization later: convert mark join to semi join, https://github.com/datafuselabs/databend/issues/7837